shell/platform/android/io/flutter/embedding/engine/android/AndroidTouchProcessor.java

```diff
@@ -57,27 +79,43 @@ public boolean onTouchEvent(MotionEvent event) {
 
     // Prepare a data packet of the appropriate size and order.
     ByteBuffer packet = ByteBuffer.allocateDirect(
-        pointerCount * POINTER_DATA_FIELD_COUNT * BYTE_PER_FIELD
+        pointerCount * POINTER_DATA_FIELD_COUNT * BYTES_PER_FIELD
     );
     packet.order(ByteOrder.LITTLE_ENDIAN);
 
     int maskedAction = event.getActionMasked();
-    // ACTION_UP, ACTION_POINTER_UP, ACTION_DOWN, and ACTION_POINTER_DOWN
-    // only apply to a single pointer, other events apply to all pointers.
-    if (maskedAction == MotionEvent.ACTION_UP || maskedAction == MotionEvent.ACTION_POINTER_UP
-        || maskedAction == MotionEvent.ACTION_DOWN || maskedAction == MotionEvent.ACTION_POINTER_DOWN) {
-      addPointerForIndex(event, event.getActionIndex(), packet);
+    int pointerChange = getPointerChangeForAction(event.getActionMasked());
+    if (maskedAction == MotionEvent.ACTION_DOWN || maskedAction == MotionEvent.ACTION_POINTER_DOWN) {
+      // ACTION_DOWN and ACTION_POINTER_DOWN always apply to a single pointer only.
+      addPointerForIndex(event, event.getActionIndex(), pointerChange, 0, packet);
+    } else if (maskedAction == MotionEvent.ACTION_UP || maskedAction == MotionEvent.ACTION_POINTER_UP) {
+      // ACTION_UP and ACTION_POINTER_UP may contain position updates for other pointers.
+      // We are converting these updates to move events here in order to preserve this data.
+      // We also mark these events with a flag in order to help the framework reassemble
+      // the original Android event later, should it need to forward it to a PlatformView.
+      for (int p = 0; p < pointerCount; p++) {
+        if (p != event.getActionIndex()) {
```

Comment:
nit, optional: These nested blocks could be flattened.

```java
if (p != event.getActionIndex() && event.getToolType(p) == MotionEvent.TOOL_TYPE_FINGER) {
```

Reply:
Done.